csrc/id_model/predicate_indexing.cpp

```diff
@@ -21,7 +21,10 @@ std::vector<IterDomain*> getPredicateDomains(
   // domains need to be predicated. Note that the non-divisible split
   // info does not seem to cover non-divisible reduction rfactor
   // splits.
-  std::vector<IterDomain*> predicate_domains = consumer_tv->hasReduction()
+  // NOTE: PreprocessGroupedMatmulInputSf's logical domain is specialized to
+  // handle padding logic and shouldn't be used for predicate.
+  std::vector<IterDomain*> predicate_domains =
+      consumer_tv->hasReduction() || expr->isA<PreprocessGroupedMatmulInputSf>()
```

Comment:
@naoyam one of the reason I added a root domain is for predicate. I know this looks hacky and blurs the boundary between root vs logical domain. Let me know if you have any strong opinion on this one.

Reply:
Is this really necessary for this PR? I suspect it's necessary in the follow-up PRs, but probably not yet here.

Reply:
It's not necessary to have it here. (I moved it from the third PR here for an early feedback).

> I thought we do not generate any code for this op but just use a runtime function.

Runtime function only handles indexing on output. Indexing on inputs are still done via codegen. (There's also a fake indexing on output to get the offset of the tensor, which I have refactored.... I can go back and try if this change is still necessary).


Nevertheless, I think the AI for this item is to revert the change and defer it to a later point.

Reply:
I see. 

Seems like resize may be useful here. Use the produce logical as the root and resize it to the logical domain. This way, predication should just work. That said, since this is a WAR, it probably wouldn't make much sense to try to polish the design.

Reply:
Good call!

I'll try that on the last codegen PR to see if it would rid this ugly patch. :saluting_face: